apps/condo/domains/common/utils/serverSchema/accessControl.js

```diff
@@ -0,0 +1,147 @@
+
+function getListPermissions (user, listKey) {
+    return transformCRUDString(get(user, ['permissions', 'lists', listKey, 'access'], ''))
+}
+
+function getFieldPermissions (user, listKey, field) {
+    return transformCRUDString(get(user, ['permissions', 'lists', listKey, 'fields', field], ''))
+}
+
+function getCustomPermissions (user, key) {
+    return get(user, ['permissions', 'custom', key], null)
+}
+
+function hasDefinedPermissions (user) {
+    return get(user, 'permissions') != null
+}
+
+function denyIfHasNoPermission (user, permission, originalAccessResult) {
+    if (originalAccessResult === false) return false
+    if (hasDefinedPermissions(user)) {
+        return permission === true
+    }
+    return originalAccessResult
+}
+
+function createStaticAccessWrapper (permissionsGetter, defaultAccessValue = true, useOperationIndexing = true) {
+    return function ({ authentication: { item: user }, operation }) {
+        const permissions = permissionsGetter(...arguments)
+        if (!useOperationIndexing) return denyIfHasNoPermission(user, permissions, defaultAccessValue) 
+        return denyIfHasNoPermission(user, permissions[operation], defaultAccessValue) 
+    }
+}
+
+function createDynamicAccessWrapper (permissionsGetter, originalAccessFn, useOperationIndexing = true){
+    return async function ({ authentication: { item: user }, operation }) {
+        const permissions = permissionsGetter(...arguments)
+        arguments[0].permissions = permissions
+        const originalAccessResult = await originalAccessFn(...arguments)
+        if (!useOperationIndexing) return denyIfHasNoPermission(user, permissions, originalAccessResult)
+        return denyIfHasNoPermission(user, permissions[operation], originalAccessResult) 
+    }
+}
+
+function perListAccess (schemaType, schemaName, schema) {
+    if (schemaType !== GQL_LIST_SCHEMA_TYPE) return schema
+
+    const { keystone } = getSchemaCtx(schemaName)
+    schema.access = parseListAccess({
+        listKey: schemaName, 
+        defaultAccess: keystone.defaultAccess.list,
+        schemaNames: ['public'],
+        access: schema.access,
+    }).public
+
+    Object.keys(schema.access).forEach(operation => {
+        let newAccFn
+        // schema.access: true
+        if (schema.access[operation] === true) {
+            const permissionsGetter = ({ authentication: { item: user }, listKey }) => getListPermissions(user, listKey)
+            newAccFn = createStaticAccessWrapper(permissionsGetter, true)
+        }
+        // schema.access[create/read/update/delete]: () => boolean
+        else if (typeof schema.access[operation] === 'function') {
+            const originalAccessFn = schema.access[operation]
+            const permissionsGetter = ({ authentication: { item: user }, listKey }) => getListPermissions(user, listKey)
+            newAccFn = createDynamicAccessWrapper(permissionsGetter, originalAccessFn, true)
+        }
+        if (newAccFn) {
```

Comment:
an unexpected case where the `schema.access` is undefined

Reply:
throw an exception

Reply:
It should be thrown by parse[List/Field/Custom] function, I will add tests for it